torch_xla/csrc/runtime/distributed_runtime.h

```diff
@@ -3,33 +3,78 @@
 
 #include <memory>
 
+#include "torch_xla/csrc/runtime/debug_macros.h"
+#include "tsl/distributed_runtime/preemption/preemption_sync_manager.h"
 #include "xla/pjrt/distributed/distributed.h"
 
 namespace torch_xla {
 namespace runtime {
 
+// DistributedRuntime serves as the point of entry for all operations which
+// required the XLA distributed runtime, such as preemption coordination.
 class DistributedRuntime {
  public:
-  static const std::string default_coordinator_port;
-  static DistributedRuntime& getInstance(int global_rank,
-                                         std::string master_addr,
-                                         std::string port) {
-    static DistributedRuntime dist_runtime_instance(global_rank, master_addr,
-                                                    port);
-    return dist_runtime_instance;
+  static inline const std::string kDefaultCoordinatorPort = "8547";
+
+  // Returns true if the distributed runtime has already been initialized.
+  static bool IsInitialized() { return dist_runtime_ != nullptr; }
+
+  // Initialize the shared DistributedRuntime object. This creates a
+  // DistributedRuntimeClient on each worker, and on global_rank 0 initializes
+  // the corresponding DistributedRuntimeService.
+  static void Initialize(int global_rank, int world_size,
+                         std::string master_addr, std::string port) {
+    XLA_CHECK(!IsInitialized()) << "DistributedRuntime already initialized";
+    dist_runtime_ = std::unique_ptr<DistributedRuntime>(
```

Comment:
umm, previously, the reason why I made `dist_runtime_` a static variable is that it guarantees that
- it only initialize only once
- it's initialized in a thread safe manner under c++11 ([so](https://stackoverflow.com/questions/246564/what-is-the-lifetime-of-a-static-variable-in-a-c-function))
- it runs its destructor even after an exception is thrown ([so](https://stackoverflow.com/questions/6418006/are-static-objects-deleted-when-an-exception-is-thrown-or-just-local-objects)) so we can make sure the distributed runtime service is shut down eventually

(3) is the most important imo.

With the current implementation, I wonder when `dist_runtime_` is destroyed.

Reply:
`dist_runtime_` is still static, just at the class level. All three should still apply, since `dist_runtime_` still has static [storage duration](https://en.cppreference.com/w/cpp/language/storage_duration).

Reply:
gotcha, thanks for pointing it out!